File: pkg/http/middleware.go
/*
 * Copyright 2025 Carver Automation Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Package http pkg/http/middleware.go
package http

import (
	"net/http"
	"strings"

	"github.com/carverauto/serviceradar/pkg/logger"
	"github.com/carverauto/serviceradar/pkg/models"
)

// CommonMiddleware handles CORS and other common HTTP concerns.
func CommonMiddleware(next http.Handler, corsConfig models.CORSConfig, log logger.Logger) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		origin := r.Header.Get("Origin")

		// If there's no Origin header, this isn't a CORS request - let it through
		if origin == "" {
			next.ServeHTTP(w, r)

			return
		}

		allowed := false

		// Check if the request origin is in the allowed list
		for _, allowedOrigin := range corsConfig.AllowedOrigins {
			if allowedOrigin == origin || allowedOrigin == "*" {
				allowed = true

				w.Header().Set("Access-Control-Allow-Origin", origin)

				break
			}
		}

		if !allowed {
			log.Warn().
				Str("origin", origin).
				Interface("allowed_origins", corsConfig.AllowedOrigins).
				Msg("CORS: Origin not allowed")
			http.Error(w, "Origin not allowed", http.StatusForbidden)

			return
		}

		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")

		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization, X-API-Key")

		w.Header().Set("Access-Control-Max-Age", "3600") // Cache preflight for 1 hour

		if corsConfig.AllowCredentials {
			w.Header().Set("Access-Control-Allow-Credentials", "true")
		} else {
			w.Header().Set("Access-Control-Allow-Credentials", "false")
		}

		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusOK)

			return
		}

		next.ServeHTTP(w, r)
	})
}

type APIKeyOptions struct {
	// API key to validate against
	APIKey string
	// Paths to exclude from API key authentication (prefix-based)
	ExcludePaths []string
	// Whether to log unauthorized attempts
	LogUnauthorized bool
	// Logger instance
	Logger logger.Logger
}

// NewAPIKeyOptions creates a new options struct with defaults.
func NewAPIKeyOptions(apiKey string, log logger.Logger) APIKeyOptions {
	return APIKeyOptions{
		APIKey:          apiKey,
		ExcludePaths:    []string{"/swagger/", "/api-docs"},
		LogUnauthorized: true,
		Logger:          log,
	}
}

// APIKeyMiddleware checks for a valid API key on requests
// excludes specified paths from authentication.
func APIKeyMiddleware(apiKey string, log logger.Logger) func(next http.Handler) http.Handler {
	opts := NewAPIKeyOptions(apiKey, log)

	return APIKeyMiddlewareWithOptions(opts)
}

// APIKeyMiddlewareWithOptions is an enhanced version of APIKeyMiddleware
// that allows configuring exclude paths and other options.
func APIKeyMiddlewareWithOptions(opts APIKeyOptions) func(next http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Check if path should be excluded
			path := r.URL.Path

			for _, excludePath := range opts.ExcludePaths {
				if strings.HasPrefix(path, excludePath) {
					// Path is excluded from authentication, let it through
					next.ServeHTTP(w, r)

					return
				}
			}

			// Check for API key
			requestKey := r.Header.Get("X-API-Key")
			if requestKey == "" {
				requestKey = r.URL.Query().Get("api_key")
			}

			if requestKey == "" || (opts.APIKey != "" && requestKey != opts.APIKey) {
				if opts.LogUnauthorized && opts.Logger != nil {
					opts.Logger.Warn().
						Str("method", r.Method).
						Str("path", r.URL.Path).
						Str("remote_addr", r.RemoteAddr).
						Msg("Unauthorized API access attempt")
				}

				http.Error(w, "Unauthorized", http.StatusUnauthorized)

				return
			}

			next.ServeHTTP(w, r)
		})
	}
}

File: web/src/middleware.ts
/*
 * Copyright 2025 Carver Automation Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// web/src/middleware.ts
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";
import { env } from "next-runtime-env";

export async function middleware(request: NextRequest) {
  const apiKey = env("API_KEY") || "";
  const isAuthEnabled = env("AUTH_ENABLED") === "true";
  const requestHeaders = new Headers(request.headers);

  // Handle OPTIONS preflight
  if (request.method === "OPTIONS") {
    return new NextResponse(null, {
      status: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET, POST, PUT, DELETE, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type, Authorization, X-API-Key",
      },
    });
  }

  // Define public paths that don't need authentication
  const publicPaths = [
    "/login",
    "/auth",
    "/serviceRadar.svg",
    "/favicons",
    "/_next",  // Next.js assets
    "/api/auth" // Authentication API routes
  ];

  // Check if the request path is public
  const isPublicPath = publicPaths.some(path => request.nextUrl.pathname.startsWith(path));
  if (isPublicPath) {
    return NextResponse.next();
  }

  // --- Main Logic for Authenticated Routes ---

  const accessToken = request.cookies.get("accessToken")?.value;
  const requestApiKey = request.headers.get("x-api-key");

  // For API routes, we allow either a bearer token or a valid API key
  if (request.nextUrl.pathname.startsWith("/api/")) {
    // Option 1: Valid Bearer Token (for UI-driven requests)
    if (accessToken) {
      requestHeaders.set("Authorization", `Bearer ${accessToken}`);
      // Also forward the API key for services that might need it internally
      if (apiKey) {
        requestHeaders.set("X-API-Key", apiKey);
      }
      return NextResponse.next({ request: { headers: requestHeaders } });
    }

    // Option 2: Valid API Key (for server-to-server requests like serviceradar-sync)
    if (requestApiKey && requestApiKey === apiKey) {
      // The API key is valid. We let the request through.
      // We also set the X-API-Key header to ensure the backend receives it.
      requestHeaders.set("X-API-Key", requestApiKey);
      return NextResponse.next({ request: { headers: requestHeaders } });
    }

    // If we reach here, it's an API request with no valid credentials.
    // Only return 401 if authentication is globally enabled.
    if (isAuthEnabled) {
      return NextResponse.json({ error: "Authentication required" }, { status: 401 });
    }
  }

  // For all other protected paths (i.e., the web UI pages), we require a bearer token if auth is enabled.
  if (isAuthEnabled) {
    if (!accessToken) {
      // No token, redirect to the login page.
      return NextResponse.redirect(new URL("/login", request.url));
    }
    // Token exists, add it to the request headers for downstream use.
    requestHeaders.set("Authorization", `Bearer ${accessToken}`);
  } else if (apiKey) {
    // If auth is disabled, fall back to using the system-wide API key for non-page requests.
    requestHeaders.set("X-API-Key", apiKey);
  }

  return NextResponse.next({
    request: { headers: requestHeaders },
  });
}

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with a specific set of prefixes.
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * - api/auth (authentication routes to prevent auth loops)
     */
    '/((?!api/auth|_next/static|_next/image|favicon.ico).*)',
  ],
};